File: Day-44/PalindromeList.java
import java.util.*;
public class PalindromeList {

  public class ListNode {
      int val;
      ListNode next;
      ListNode() {}
      ListNode(int val) { this.val = val; }
      ListNode(int val, ListNode next) { this.val = val; this.next = next; }
  }
 

    public boolean isPalindrome(ListNode head) {
       Stack<Integer> st=new Stack<>();
       ListNode temp=head;
       while(temp!=null)
       {
        st.push(temp.val);
        temp=temp.next;
       }
       temp=head;
       while(temp!=null)
       {
        if(temp.val!=st.peek()) return false;
        st.pop();
        temp=temp.next;
       }
       return true;
    }
}

//TC:0(2N)
//SC:O(N) for stack

File: Day-44/PalindromeList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        st=[]
        temp=head
        while temp:
            st.append(temp.val)
            temp=temp.next
        temp=head
        while temp:
            if st.pop() != temp.val:
                return False
            #st.pop()
            temp=temp.next
        return True

        